### PR TITLE
makes backoff factory configurable

### PR DESCRIPTION
**Pull request by xh3b4sd**
Fixes https://github.com/giantswarm/giantswarm/issues/4031.

### CODE REVIEW

File: resource.go [hunk 1]
```diff
@@ -58,8 +62,8 @@ type Resource struct {
 }
 
 func New(config Config) (*Resource, error) {
-	if config.ClusterEndpointFunc == nil {
```
Comment (kopiczko) [inline resource.go]: What happened to the `ClusterEndpointFunc` check? Is diff weird?

File: resource.go [hunk 2]
```diff
@@ -20,6 +22,7 @@ const (
 )
 
 type Config struct {
+	BackOffFactory      func() backoff.Interface
```
Comment (kopiczko) [inline resource.go]: Do we even need to configure that? Maybe an internal is a good fit here?

File: resource.go [hunk 3]
```diff
@@ -58,8 +62,8 @@ type Resource struct {
 }
 
 func New(config Config) (*Resource, error) {
-	if config.ClusterEndpointFunc == nil {
-		return nil, microerror.Maskf(invalidConfigError, "%T.ClusterEndpointFunc must not be empty", config)
+	if config.BackOffFactory == nil {
+		config.BackOffFactory = func() backoff.Interface { return backoff.NewMaxRetries(3, 1*time.Second) }
```
Comment (MarcelMue) [inline resource.go]: I am a bit confused how long we want to retry here. This seems a little short now but no strong feelings.

**Review by MarcelMue — APPROVED** [review]

**Review by kopiczko — APPROVED** [review]

File: resource.go (hunk 3 above)
Comment (kopiczko) [reply resource.go]: We have also retryresource, check the issue linked in the desc.

Comment (xh3b4sd) [reply resource.go]: This should be fine.

File: resource.go (hunk 1 above)
Comment (xh3b4sd) [reply resource.go]: I was too fast. Brought it back.

File: resource.go (hunk 2 above)
Comment (xh3b4sd) [reply resource.go]: It is configurable now. We default it. Good enough. Lets go.

**Pull request merged**